Support for `antnode` Automatic Upgrades

jacderida opened the pull request:
- 60a3fc9c2 **chore: utility target for providing binary hashes**

  The automatic upgrading process is going to require the sha256 hashes of the binaries be included in
  the information for each release.

  This new target in the Justfile can operate on the populated `artifacts` directory to get all the
  hashes. It will output them in a markdown table which can be pasted into the release description.

- c9ed698d3 **feat: `antnode` automatic upgrades for unix platforms**

  Automatic upgrades aim to make it as easy as possible for node operators to obtain new versions of
  the `antnode` binary. We also provide a 'staggering' mechanism to attempt to avoid the turbulence
  caused when everyone upgrades around the same time. If the majority of operators use this feature,
  we should see better network health and stability, because newer versions of the node will propagate
  quicker and safer.

  Initially, we only support Linux and macOS. The Windows operating system does not allow a binary
  associated with a running process to be overwritten, which is part of the upgrade process (see
  below).

  An upgrade is performed by a background thread that wakes up every 72 hours.

  After the upgrade is performed, the restart of the binary is delayed for a period of time that is
  determined by taking the hash of the peer ID and the size of the network into account, with a cap of
  72 hours. This is to make the upgrades 'staggered', to avoid many thousands of node processes
  restarting at the same time. To facilitate the calculation, the `get_kbuckets` API is slightly
  modified to return the estimated network size.

  The upgrade process works as follows:

  * Get the latest release information from the `autonomi` repository.
  * If the commit hash of the latest release matches the commit hash of the currently running binary,
    do nothing and the process ends. Otherwise the upgrade proceeds.
  * Get the currently running platform. This information is used to locate the correct binary and
    sha256 hash.
  * Check if another `antnode` process has already downloaded the new binary to a temporary location
    and its sha256 hash matches that of the binary from the fetched release information. This will
    prevent many `antnode` processes continually downloading the new binary. If the new binary has not
    already been downloaded, it will be downloaded and extracted to the temporary location.
  * The hash of the downloaded binary will be compared to the hash of the binary in the new release
    information.
  * The binary of the currently running process will be overwritten by the new binary. We will check
    to see if the hash of the running binary matches the hash of the upgrade binary; in this case,
    another process already upgraded the binary and we don't need to do anything.

  If an upgrade occurred, the `antnode` process will be restarted after a delay of some period, as
  mentioned above.

  Given the more elaborate implementation of automatic upgrades, the workflow for self restarting was
  deleted.

- a506e71c6 **chore!: use 'always' restart policy for `antctl add`**

  BREAKING CHANGE: new node deployments with `antctl` will apply a different restart policy.

  With `antctl` we are going to configure services for node deployments to always restart. This is so
  we can use the node deployments with the automatic upgrades feature and let the service manager
  handle the restart of the `antnode` process. If `antnode` itself were to restart the process, the
  service manager would consider it to be dead, which would break the status tracking for `antctl`.

  We need to provide a `--stop-on-upgrade` argument for `antnode` such that it will stop the process
  on the upgrade rather than restart it, and allow the service manager to perform the restart. This
  will be coming in the next commit.

  The `node-launchpad` has been configured to use 'never' for the restart policy because we ran into
  problems with the UPnP feature if the launchpad continually restarted nodes. So automatic upgrades
  just won't be supported when using the launchpad. We will need to provide another `antnode` argument
  to disable automatic upgrades, which will come in another commit.

- 5e5528283 **feat: provide `--stop-on-upgrade` argument for `antctl`**

  If this argument is used the node will be stopped after it is upgraded, rather than restarted.

  This is to allow the service manager to restart the process if the node is being run in this
  context, which it will be if it is deployed using `antctl`, which is updated in this commit to
  always supply the argument as part of the service definition for a node.

- 48ec66a98 **chore: improve logging for upgrades**

  I added the information I found useful during the development process.

- 609a2a2a8 **fix: do not truncate upgrade lock file**

  By examining the logs on a testnet run I determined all the `antnode` processes on the same
  host downloaded the upgrade binary. This is something we expressly wanted to avoid.

  The problem was the `File::create` function was truncating the lock file. It has now been changed to
  avoid this behaviour, so other processes should wait to acquire the lock.

- d84051e89 **chore: output node port to log**

  This will be used to verify all nodes participating in the automatic-upgrade process will be
  restarted with the same port.
  
- 03b5e9acd **feat: provide cache for release info**

  In a similar fashion to the cache for the upgrade binary, we provide a cache for the release
  information.

  While performing a test with 5000 nodes I got rate limited by Github when thousands of processes
  queried for the release information at approximately the same time. Now it should only be one query
  per host.

- ef9e8e6af **chore: use sha256 hash in restart delay calculation**

  In the test run with 1000 nodes we observed a distribution of restart times that seemed a bit too
  tightly clustered.

  Using this alternative hashing mechanism resulted in a distribution that was more even.

  The restart delay calculation is also moved outside the loop because it only needs to be calculated
  once.

- 7b89cac04 **chore: more elaborate error handling**

  An `UpgradeError` type is introduced to make it easier to distinguish between errors and the case
  where there is no new release available. The latter of those will now log at `INFO` level.

jacderida committed:
chore: utility target for providing binary hashes

The automatic upgrading process is going to require the sha256 hashes of the binaries be included in
the information for each release.

This new target in the Justfile can operate on the populated `artifacts` directory to get all the
hashes. It will output them in a markdown table which can be pasted into the release description.

jacderida committed:
feat: `antnode` automatic upgrades for unix platforms

Automatic upgrades aim to make it as easy as possible for node operators to obtain new versions of
the `antnode` binary. We also provide a 'staggering' mechanism to attempt to avoid the turbulence
caused when everyone upgrades around the same time. If the majority of operators use this feature,
we should see better network health and stability, because newer versions of the node will propagate
quicker and safer.

Initially, we only support Linux and macOS. The Windows operating system does not allow a binary
associated with a running process to be overwritten, which is part of the upgrade process (see
below).

An upgrade is performed by a background thread that wakes up every 72 hours.

After the upgrade is performed, the restart of the binary is delayed for a period of time that is
determined by taking the hash of the peer ID and the size of the network into account, with a cap of
72 hours. This is to make the upgrades 'staggered', to avoid many thousands of node processes
restarting at the same time. To facilitate the calculation, the `get_kbuckets` API is slightly
modified to return the estimated network size.

The upgrade process works as follows:

* Get the latest release information from the `autonomi` repository.
* If the commit hash of the latest release matches the commit hash of the currently running binary,
  do nothing and the process ends. Otherwise the upgrade proceeds.
* Get the currently running platform. This information is used to locate the correct binary and
  sha256 hash.
* Check if another `antnode` process has already downloaded the new binary to a temporary location
  and its sha256 hash matches that of the binary from the fetched release information. This will
  prevent many `antnode` processes continually downloading the new binary. If the new binary has not
  already been downloaded, it will be downloaded and extracted to the temporary location.
* The hash of the downloaded binary will be compared to the hash of the binary in the new release
  information.
* The binary of the currently running process will be overwritten by the new binary. We will check
  to see if the hash of the running binary matches the hash of the upgrade binary; in this case,
  another process already upgraded the binary and we don't need to do anything.

If an upgrade occurred, the `antnode` process will be restarted after a delay of some period, as
mentioned above.

Given the more elaborate implementation of automatic upgrades, the workflow for self restarting was
deleted.

jacderida committed:
chore!: use 'always' restart policy for `antctl add`

BREAKING CHANGE: new node deployments with `antctl` will apply a different restart policy.

With `antctl` we are going to configure services for node deployments to always restart. This is so
we can use the node deployments with the automatic upgrades feature and let the service manager
handle the restart of the `antnode` process. If `antnode` itself were to restart the process, the
service manager would consider it to be dead, which would break the status tracking for `antctl`.

We need to provide a `--stop-on-upgrade` argument for `antnode` such that it will stop the process
on the upgrade rather than restart it, and allow the service manager to perform the restart. This
will be coming in the next commit.

The `node-launchpad` has been configured to use 'never' for the restart policy because we ran into
problems with the UPnP feature if the launchpad continually restarted nodes. So automatic upgrades
just won't be supported when using the launchpad. We will need to provide another `antnode` argument
to disable automatic upgrades, which will come in another commit.

jacderida committed:
feat: provide `--stop-on-upgrade` argument for `antctl`

If this argument is used the node will be stopped after it is upgraded, rather than restarted.

This is to allow the service manager to restart the process if the node is being run in this
context, which it will be if it is deployed using `antctl`, which is updated in this commit to
always supply the argument as part of the service definition for a node.

jacderida committed:
chore: improve logging for upgrades

I added the information I found useful during the development process.

jacderida committed:
fix: do not truncate upgrade lock file

By examining the logs on a testnet run I determined all the `antnode` processes on the same
host downloaded the upgrade binary. This is something we expressly wanted to avoid.

The problem was the `File::create` function was truncating the lock file. It has now been changed to
avoid this behaviour, so other processes should wait to acquire the lock.

jacderida committed:
chore: output node port to log

This will be used to verify all nodes participating in the automatic-upgrade process will be
restarted with the same port.

jacderida committed:
feat: provide cache for release info

In a similar fashion to the cache for the upgrade binary, we provide a cache for the release
information.

While performing a test with 5000 nodes I got rate limited by Github when thousands of processes
queried for the release information at approximately the same time. Now it should only be one query
per host.

maqi commented on ant-node/src/bin/antnode/main.rs:
```diff
@@ -422,17 +436,71 @@ You can check your reward balance by running:
+                // 72 hours (259200 seconds) with ±5% randomization to prevent simultaneous upgrades
+                let base_delay = 259200;
+                let variance = rand::thread_rng().gen_range(-12960..=12960);
+                let upgrade_check_delay_secs = (base_delay + variance) as u64;
```
this block of `upgrade_check_delay_secs` calculatin can be outside of the loop

maqi commented on ant-node/src/bin/antnode/main.rs:
```diff
@@ -422,17 +436,71 @@ You can check your reward balance by running:
+                        break;
+                    }
+                    Err(e) => {
+                        error!("Error during upgrade process: {e}");
```
`Not to perform an upgrade` shall not be counted as an `error`, as it happens for most of the time ?

jacderida replied:
An `UpgradeError` has been introduced to distinguish this scenario.

maqi commented on ant-node/Cargo.toml:
```diff
@@ -28,6 +28,7 @@ ant-build-info = { path = "../ant-build-info", version = "0.1.29" }
 ant-evm = { path = "../ant-evm", version = "0.1.17" }
 ant-logging = { path = "../ant-logging", version = "0.3.0", features = ["process-metrics"] }
 ant-protocol = { path = "../ant-protocol", version = "1.0.9" }
+ant-releases = { git = "https://github.com/jacderida/ant-releases", branch = "chore-serialization_for_release_info" }
```
just a note that his shall be updated before merge

maqi commented on ant-node-manager/Cargo.toml:
```diff
@@ -51,7 +51,7 @@ semver = "1.0.20"
 serde_json = "1.0"
 reqwest = { version = "0.12", default-features = false, features = ["json", "rustls-tls"] }
-service-manager = "0.8.0"
+service-manager = { git = "https://github.com/jacderida/service-manager-rs", branch = "generic_restart_policy" }
```
just a note that his shall be updated before merge

jacderida replied:
Yeah, the change for this crate is going to need to be reviewed, so I will merge this PR into the `automatic_upgrades` branch with the Git-based reference, but it will be resolved before going back to `main`.

maqi commented on ant-node-manager/Cargo.toml:
```diff
@@ -34,7 +34,7 @@ ant-build-info = { path = "../ant-build-info", version = "0.1.29" }
 ant-logging = { path = "../ant-logging", version = "0.3.0" }
 ant-protocol = { path = "../ant-protocol", version = "1.0.9" }
-ant-releases = { version = "0.4.1" }
+ant-releases = { version = "0.4.2" }
```
just to confirm, so here doesn't need to use the same bespoken branch as ant-node ?

jacderida replied:
It's easier to just keep all the `ant-release` references on the same version.

maqi commented on ant-service-management/Cargo.toml:
```diff
@@ -22,7 +22,7 @@ prost = { version = "0.9" }
 serde_json = "1.0"
 semver = "1.0.20"
-service-manager = "0.8.0"
+service-manager = { git = "https://github.com/jacderida/service-manager-rs", branch = "generic_restart_policy" }
```
just a note that his shall be updated before merge

jacderida committed:
chore: use sha256 hash in restart delay calculation

In the test run with 1000 nodes we observed a distribution of restart times that seemed a bit too
tightly clustered.

Using this alternative hashing mechanism resulted in a distribution that was more even.

The restart delay calculation is also moved outside the loop because it only needs to be calculated
once.

jacderida committed:
chore: more elaborate error handling

An `UpgradeError` type is introduced to make it easier to distinguish between errors and the case
where there is no new release available. The latter of those will now log at `INFO` level.
